File: app/evolution_utils.py
"""Evolution chain utilities — Python mirror of lib/evolution-utils.ts."""

GENERATION_ORDER = [
    "generation-i", "generation-ii", "generation-iii", "generation-iv",
    "generation-v", "generation-vi", "generation-vii", "generation-viii",
    "generation-ix",
]

GENERATION_MAX_ID = {
    "generation-i": 151, "generation-ii": 251, "generation-iii": 386,
    "generation-iv": 493, "generation-v": 649, "generation-vi": 721,
    "generation-vii": 809, "generation-viii": 905, "generation-ix": 1025,
}


def generation_index(gen: str) -> int:
    try:
        return GENERATION_ORDER.index(gen)
    except ValueError:
        return -1


def extract_id_from_url(url: str) -> int:
    return int(url.rstrip("/").split("/")[-1])


def flatten_chain(link: dict) -> dict:
    return {
        "name": link["species"]["name"],
        "id": extract_id_from_url(link["species"]["url"]),
        "evolution_details": link.get("evolution_details", []),
        "evolves_to": [flatten_chain(c) for c in link.get("evolves_to", [])],
    }


def _capitalize(s: str) -> str:
    return s[0].upper() + s[1:] if s else s


def _format_name(name: str) -> str:
    return " ".join(_capitalize(p) for p in name.split("-"))


def _pick_detail_for_generation(details: list[dict], generation: str) -> dict | None:
    if not details:
        return None
    if len(details) == 1:
        return details[0]

    gen_idx = generation_index(generation)
    has_location = any(d.get("location") is not None for d in details)
    has_item = any(
        d.get("item") is not None and d.get("trigger", {}).get("name") == "use-item"
        for d in details
    )

    if has_location and has_item:
        if gen_idx >= generation_index("generation-viii"):
            return next(
                (d for d in details if d.get("item") and d.get("trigger", {}).get("name") == "use-item"),
                details[0],
            )
        return next((d for d in details if d.get("location")), details[0])

    return details[0]


def _format_single_detail(d: dict) -> str:
    parts: list[str] = []
    trigger = d.get("trigger", {}) or {}
    trigger_name = trigger.get("name", "")

    if trigger_name == "trade":
        held = d.get("held_item")
        if held:
            parts.append(f"Trade ({_format_name(held['name'])})")
        else:
            parts.append("Trade")
    elif trigger_name == "use-item" and d.get("item"):
        parts.append(_format_name(d["item"]["name"]))
    elif d.get("min_level"):
        parts.append(f"Lv. {d['min_level']}")

    if d.get("min_happiness"):
        parts.append(f"Happiness ≥ {d['min_happiness']}")
    if d.get("time_of_day"):
        parts.append(f"at {_capitalize(d['time_of_day'])}")
    if d.get("known_move_type"):
        parts.append(f"Know {_capitalize(d['known_move_type']['name'])} move")
    if d.get("known_move"):
        parts.append(f"Know {_format_name(d['known_move']['name'])}")
    if d.get("location") and trigger_name != "use-item":
        parts.append(f"at {_format_name(d['location']['name'])}")
    if d.get("min_beauty"):
        parts.append(f"Beauty ≥ {d['min_beauty']}")
    if d.get("min_affection"):
        parts.append(f"Affection ≥ {d['min_affection']}")
    if d.get("gender") is not None:
        parts.append("♀ only" if d["gender"] == 1 else "♂ only")
    if d.get("needs_overworld_rain"):
        parts.append("in Rain")
    if d.get("turn_upside_down"):
        parts.append("Upside Down")
    if d.get("relative_physical_stats") is not None:
        rps = d["relative_physical_stats"]
        if rps == 1:
            parts.append("Atk > Def")
        elif rps == -1:
            parts.append("Atk < Def")
        else:
            parts.append("Atk = Def")

    if not parts and trigger_name:
        parts.append(_format_name(trigger_name))

    return ", ".join(parts)


def format_evolution_method(details: list[dict], generation: str) -> str:
    if not details:
        return ""
    picked = _pick_detail_for_generation(details, generation)
    if not picked:
        return ""
    return _format_single_detail(picked)


def is_available_in_generation(pokemon_id: int, generation: str) -> bool:
    max_id = GENERATION_MAX_ID.get(generation, 0)
    return pokemon_id <= max_id

File: lib/evolution-utils.ts
import type { EvolutionChainLink, EvolutionDetail } from "./pokeapi";
import { extractIdFromUrl } from "./pokeapi";
import { GENERATION_MAX_ID, generationIndex, type GenerationName } from "./constants";

export interface EvolutionStage {
  name: string;
  id: number;
  evolutionDetails: EvolutionDetail[];
  evolvesTo: EvolutionStage[];
}

export function flattenChain(link: EvolutionChainLink): EvolutionStage {
  return {
    name: link.species.name,
    id: extractIdFromUrl(link.species.url),
    evolutionDetails: link.evolution_details,
    evolvesTo: link.evolves_to.map(flattenChain),
  };
}

/**
 * Format an evolution method for display, picking the generation-appropriate
 * entry when multiple methods exist.
 */
export function formatEvolutionMethod(
  details: EvolutionDetail[],
  generation: GenerationName,
): string {
  if (details.length === 0) return "";

  const picked = pickDetailForGeneration(details, generation);
  if (!picked) return "";

  return formatSingleDetail(picked);
}

function pickDetailForGeneration(
  details: EvolutionDetail[],
  generation: GenerationName,
): EvolutionDetail | null {
  if (details.length === 1) return details[0];

  const genIdx = generationIndex(generation);
  const hasLocation = details.some((d) => d.location !== null);
  const hasItem = details.some((d) => d.item !== null && d.trigger?.name === "use-item");

  // Leafeon/Glaceon heuristic: location-based for Gen IV-VII, item-based for Gen VIII+
  if (hasLocation && hasItem) {
    if (genIdx >= generationIndex("generation-viii")) {
      return details.find((d) => d.item !== null && d.trigger?.name === "use-item") ?? details[0];
    }
    return details.find((d) => d.location !== null) ?? details[0];
  }

  return details[0];
}

function formatSingleDetail(d: EvolutionDetail): string {
  const parts: string[] = [];

  if (d.trigger?.name === "trade") {
    if (d.held_item) {
      parts.push(`Trade (${formatName(d.held_item.name)})`);
    } else {
      parts.push("Trade");
    }
  } else if (d.trigger?.name === "use-item" && d.item) {
    parts.push(formatName(d.item.name));
  } else if (d.min_level) {
    parts.push(`Lv. ${d.min_level}`);
  }

  if (d.min_happiness) {
    parts.push(`Happiness ≥ ${d.min_happiness}`);
  }

  if (d.time_of_day && d.time_of_day !== "") {
    parts.push(`at ${capitalize(d.time_of_day)}`);
  }

  if (d.known_move_type) {
    parts.push(`Know ${capitalize(d.known_move_type.name)} move`);
  }

  if (d.known_move) {
    parts.push(`Know ${formatName(d.known_move.name)}`);
  }

  if (d.location && d.trigger?.name !== "use-item") {
    parts.push(`at ${formatName(d.location.name)}`);
  }

  if (d.min_beauty) {
    parts.push(`Beauty ≥ ${d.min_beauty}`);
  }

  if (d.min_affection) {
    parts.push(`Affection ≥ ${d.min_affection}`);
  }

  if (d.gender !== null) {
    parts.push(d.gender === 1 ? "♀ only" : "♂ only");
  }

  if (d.needs_overworld_rain) {
    parts.push("in Rain");
  }

  if (d.turn_upside_down) {
    parts.push("Upside Down");
  }

  if (d.relative_physical_stats !== null) {
    if (d.relative_physical_stats === 1) parts.push("Atk > Def");
    else if (d.relative_physical_stats === -1) parts.push("Atk < Def");
    else parts.push("Atk = Def");
  }

  if (parts.length === 0 && d.trigger) {
    parts.push(formatName(d.trigger.name));
  }

  return parts.join(", ");
}

function formatName(name: string): string {
  return name.split("-").map(capitalize).join(" ");
}

function capitalize(s: string): string {
  return s.charAt(0).toUpperCase() + s.slice(1);
}

/**
 * Determine if a Pokémon (by national dex ID) existed in a given generation.
 */
export function isAvailableInGeneration(pokemonId: number, generation: GenerationName): boolean {
  return pokemonId <= GENERATION_MAX_ID[generation];
}

/**
 * Get the sprite URL for an evolution stage without a full API call.
 */
export function getEvolutionSpriteUrl(pokemonId: number): string {
  return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${pokemonId}.png`;
}
